### PR TITLE
Socketio metadata

### PR DESCRIPTION
**Pull request by rgstephens**
**Proposed changes**:
- Adds metadata to the socket.io channel as requested in #11087 

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/core/channels/socketio.py
Comment (ancalita) [inline rasa/core/channels/socketio.py:169]: Why is this new attribute necessary?

File: rasa/core/channels/socketio.py
Comment (ancalita) [inline rasa/core/channels/socketio.py:159]: If we are passing the value that corresponds to `metadata_key` in `credentials` dict, then that can be of type `Optional[Any]`? If it must be `Optional[Text]` the docs need to add this clarification, because the example uses `customData` and that threw me off a bit 😅

**Review by ancalita — COMMENTED** [review]
Apologies for the delay 🙈 
Looks almost ready, I asked for a few clarifications.

File: rasa/core/channels/socketio.py
Comment (rgstephens) [reply rasa/core/channels/socketio.py:169]: Removing `environ`. (I was looking at adding an option to channel to include select environment values that could be useful as metadata.)

File: rasa/core/channels/socketio.py
Comment (rgstephens) [reply rasa/core/channels/socketio.py:208]: Will remove these two lines

File: rasa/core/channels/socketio.py
Comment (rgstephens) [reply rasa/core/channels/socketio.py:159]: I don't understand. The `metadata_key` value would need to be text.

Comment (ancalita) [reply rasa/core/channels/socketio.py:159]: I got confused by the docs yaml example `metadata_key: customData` and expected the value in this pair to be the actual metadata object or dictionary. But I think you actually want this to be a string key, so I'd amend the example in the docs with `metadata_key: your_custom_metadata_key`

Comment (rgstephens) [reply rasa/core/channels/socketio.py:159]: Thanks, I chose `customData` because the most widely used rasa chat widget, [botfront widget](https://github.com/botfront/rasa-webchat#rasa) stores metadata under this key.

**Comment by ancalita** [comment]
@rgstephens Following up on this if you'd want a 2nd review from me?

**Comment by rgstephens** [comment]
yes, thanks for the reminder

File: docs/docs/connectors/your-own-website.mdx
Comment (ancalita) [inline docs/docs/connectors/your-own-website.mdx:119]: Another clarifying question - sorry, looking at this with fresh eyes after a while 😄 - What does `another object name` refer to in this sentence "The `metadata_key` setting can be used to supply metadata on another object name."

**Review by ancalita — COMMENTED** [review]
Before I give the final approval, asked a clarifying question on a docs sentence. Also please add a unit test if possible, that the metadata gets added to the `UserMessage` object.

File: docs/docs/connectors/your-own-website.mdx
Comment (ancalita) [reply docs/docs/connectors/your-own-website.mdx:119]: Sounds good 👍🏻

Comment (rgstephens) [reply docs/docs/connectors/your-own-website.mdx]: pushed updated text and resolved conflict

**Review by ancalita — CHANGES_REQUESTED** [review]
Almost there 💯 
I notice this is targeting `3.1.x`, we no longer automatically merge `3.1.x` or `3.0.x` into `main` since the latest release branch is `3.2.x`, once this is merged, would you be able please to port this change to `main` too? Alternatively cherry-picking and rebasing to `3.2.x`?

File: rasa/core/channels/socketio.py
Comment (rgstephens) [reply rasa/core/channels/socketio.py:264]: I rebased to `3.2.x`.  Regarding the test for metadata, isn't this line protecting against the metadata not being passed and setting the value an empty object?

`metadata = data.get(self.metadata_key, {})`

**Comment by rgstephens** [comment]
> Before I give the final approval, asked a clarifying question on a docs sentence. Also please add a unit test if possible, that the metadata gets added to the `UserMessage` object.

Have not done unit tests before in Python so this make take me a while. Might be best to withdraw this PR rather than waiting on me to do this.

File: rasa/core/channels/socketio.py
Comment (ancalita) [reply rasa/core/channels/socketio.py:264]: > rebased to 3.2.x.

Great, the target branch in GH has to be edited to `3.2.x` too, right now it is still `3.1.x`.

> isn't this line protecting against the metadata not being passed and setting the value an empty object?

Yup, that's right.

**Comment by ancalita** [comment]
> > Before I give the final approval, asked a clarifying question on a docs sentence. Also please add a unit test if possible, that the metadata gets added to the `UserMessage` object.
> 
> Have not done unit tests before in Python so this make take me a while. Might be best to withdraw this PR rather than waiting on me to do this.

As you prefer!

**Comment by ancalita** [comment]
@rgstephens Have you rebased recently? I just tried to review but see over 200 commits in this PR, might be easier to recreate PR targeting `3.2.x` and close this one, we are actually planning a micro by the end of this week, latest early next week. I am available to give a fast review too 👍🏻

**Comment by rgstephens** [comment]
Closing. Rebased to `3.2.x` and replaced by PR #11435

**Pull request closed without merging**
